common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -976,9 +977,28 @@ public UTF8String replace(UTF8String search, UTF8String replace) {
     if (EMPTY_UTF8.equals(search)) {
       return this;
     }
-    String replaced = toString().replace(
-      search.toString(), replace.toString());
-    return fromString(replaced);
+    return replace(search.toString(), replace.toString());
+  }
+
+  public UTF8String replace(String search, String replace) {
+    String before = toString();
+    String after;
+    if (search.length() == 1 && replace.length() == 1) {
+      // Use single-character-replacement fast path
+      after = before.replace(search.charAt(0), replace.charAt(0));
+    } else {
+      // In Java 8, String.replace() is implemented using a regex and is therefore
+      // somewhat inefficient (see https://bugs.openjdk.java.net/browse/JDK-8058779).
+      // This is fixed in Java 9, but in Java 8 we can use Commons StringUtils instead:
+      after = StringUtils.replace(before, search, replace);
+    }
+    // Use reference equality to cheaply detect whether the replacement had no effect,
+    // in which case we can simply return the original UTF8String and save some copying.
+    if (before == after) {
+      return this;
```

Comment:
One consideration here: do we need to make a defensive copy? If so, we can't do this optimization.

Why might we need to copy a `UTF8String`? The `UTF8String` instance itself is effectively immutable, but the underlying storage might be a region of potentially-not-exclusively-owned memory (either direct/off-heap memory or a region of a `long[]` array), so we might need to perform a copy in case we're going to buffer / otherwise hold onto the `UTF8String` past a point where the underlying underlying storage memory could be mutated.

I think the most common case to worry about would be a UTF8String which is backed by memory that is part of a larger `UnsafeRow`. If we're doing row-at-a-time processing and aren't holding onto this `UTF8String` across rows then I think we're ok since changes to rows' memory _during_ single-row processing would impact many parts of Spark and would probably be detected. In the few places where we _do_ hold references across evaluations / rows then we need to copy, but I suspect most places already do this: for example, see the `regexp.clone()` in the `RegExpReplace` expression.

My intuition is that we probably _don't_ need to make a defensive copy here because I doubt we have parts of the code which _specifically assume that `replace()` will copy_ (i.e. which are abusing `replace()` as a slow `clone()` mechanism). Put differently, I suspect that any code which would fail due to lack of copying in `replace()` is also vulnerable to this problem from other sources (including simply reading a string from a row without further modification), so I don't think we need to add extra copying here.

I'd love to get additional sets of eyes on this, though, and I'd ultimately be ok with changing `return this` to `return this.clone()` (and updating the other `return this` uses in `UTF8String`) if we conclude that this isn't safe (or are uncertain and want to err on the side of caution).

Reply:
I think your intuition is right here.